transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,419 @@
+use futures::{Future, SinkExt, Stream, StreamExt};
+use futures_timer::Delay;
+use std::{
+    cmp::Reverse,
+    collections::BinaryHeap,
+    future::poll_fn,
+    sync::{Arc, Mutex},
+    task::Poll,
+    time::Duration,
+};
+use tracing::trace;
+
+use crate::{config::MixerConfig, data::DelayedData};
+
+#[cfg(all(feature = "prometheus", not(test)))]
+use hopr_metrics::metrics::SimpleGauge;
+
+#[cfg(all(feature = "prometheus", not(test)))]
+lazy_static::lazy_static! {
+    static ref METRIC_QUEUE_SIZE: SimpleGauge =
+        SimpleGauge::new("hopr_mixer_queue_size", "Current mixer queue size").unwrap();
+    static ref METRIC_MIXER_AVERAGE_DELAY: SimpleGauge = SimpleGauge::new(
+        "hopr_mixer_average_packet_delay",
+        "Average mixer packet delay averaged over a packet window"
+    )
+    .unwrap();
+}
+
+/// Mixing and delaying channel using random delay function.
+///
+/// Mixing is performed by assigning random delays to the ingress timestamp of data,
+/// then storing the values inside a binary heap with reversed ordering (max heap).
+/// This effectively creates a min heap behavior, which is required to ensure that
+/// data is released in order of their delay expiration.
+///
+/// When data arrives:
+/// 1. A random delay is assigned
+/// 2. Data is stored in the heap with its release timestamp
+/// 3. The heap maintains ordering so items with earliest release time are at the top
+///
+/// The channel uses a single timer thread that is instantiated on the first
+/// timer reset and shared across all operations.
+struct Channel<T> {
```

Comment:
Should not the channel have a bound?

Reply:
What type of bound do you mean? It has capacity, but is able to rescale, at the cost of halting operations for reallocations of the buffer.

Reply:
Sure, but do we want this? I'd put a cap on the growth of the BinaryHeap, where poll_ready will be pending until the Stream side evicts sufficient number of elemetns, as an example.

Reply:
This channel was never designed to be bounded like that. All of the internal channels are unbounded, with ingress and egress being responsible for throttling the throughput.

Reply:
Right, the inputs are bounded, but the number of them is potentially unbounded as they all meet in the mixer, no? On a relay that handles multiple nodes. If there are many nodes sending little traffic, it can still blow up the BinaryHeap size.

Reply:
No, the bound applies directly on ingress, so there will not be more packets than capacity. It's the same for egress.